jlm/llvm/opt/alias-analyses/Worklist.hpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_WORKLIST_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_WORKLIST_HPP
+
+#include <jlm/util/common.hpp>
+#include <jlm/util/HashSet.hpp>
+
+#include <limits>
+#include <queue>
+#include <stack>
+#include <unordered_map>
+#include <vector>
+
+namespace jlm::llvm::aa
+{
+
+/**
+ * Class for managing a set of work items, that are waiting to be visited by an algorithm.
+ * The implementation decides in what order remaining work items should be processed.
+ * @tparam T the type of the work items.
+ */
+template<typename T>
+class Worklist
+{
+public:
+  virtual ~Worklist() = default;
+
+  Worklist() = default;
+
+  Worklist(const Worklist & other) = delete;
+
+  Worklist(Worklist && other) = delete;
+
+  Worklist &
+  operator=(const Worklist & other) = delete;
+
+  Worklist &
+  operator=(Worklist && other) = delete;
+
+  /**
+   * @return true if there are work items left to be visited
+   */
+  [[nodiscard]] virtual bool
+  HasMoreWorkItems() = 0;
+
+  /**
+   * Removes one work item from the worklist.
+   * Requires there to be at least one work item left.
+   * @return the removed work item.
+   */
+  [[nodiscard]] virtual T
+  PopWorkItem() = 0;
+
+  /**
+   * Adds a work item to the worklist.
+   * If the item is already present, the item is not added again, but its position may be changed.
+   * @param item the work item to be added.
+   */
+  virtual void
+  PushWorkItem(T item) = 0;
+};
+
+/**
+ * Worklist implementation using a stack.
+ * Pushing a work item that is already on the stack is a no-op.
+ * @tparam T the type of the work items.
+ * @see Worklist
+ */
+template<typename T>
+class LifoWorklist final : public Worklist<T>
+{
+public:
+  ~LifoWorklist() override = default;
+
+  LifoWorklist() = default;
+
+  bool
+  HasMoreWorkItems() override
+  {
+    return !WorkItems_.empty();
+  }
+
+  T
+  PopWorkItem() override
+  {
+    JLM_ASSERT(HasMoreWorkItems());
+    T item = WorkItems_.top();
+    WorkItems_.pop();
+    OnList_.Remove(item);
+    return item;
+  }
```

Comment:
Since we are copying all the types into both the OnList and the queue/stack itself, the class really shouldn't be used with types that are not trivially copyable. In that case the classes should be written to not make copies, and use stable references.